### ISSUE TITLE
[Feature package] Full feature support with Ascend NPU

### ISSUE THREAD

**Issue by hipudding** [microsoft/DeepSpeed#4567]
## Background

**Ascend** is a full-stack AI computing infrastructure for industry applications and services based on Huawei Ascend processors and software. For more information about Ascend, see [Ascend Community](https://www.hiascend.com/en/).

**[CANN](https://www.hiascend.com/en/software/cann)** (Compute Architecture of Neural Networks), developped by Huawei, is a heterogeneous computing architecture for AI.

**Pytorch** has officially announced [support for Ascend NPU](https://pytorch.org/blog/pytorch-2-1/) (through key PrivateUse1), please see the PrivateUse1 tutorial [here](https://pytorch.org/tutorials/advanced/privateuseone.html).

## Previous work

NPU accelerator support has already been merged (see #3595, #3831), which makes it possible to use NPU as a backend accelerator for basic training and inferencing tasks. However, to achieve full support, more features need to be implemented.

## Sub tasks

Here is a list of features that need to be implemented or tested.

status|title|assigned to
--|--|--
Done|[Unit tests](https://github.com/microsoft/DeepSpeed/pull/4569)|@RUAN-ZX
Done|FP16|@minchao-sun, @wuhhu
Done|BF16|@minchao-sun, @wuhhu
Done|Gradient Accumulation|@minchao-sun, @wuhhu
Done|Data Parallelism|@minchao-sun, @wuhhu
Done|Pipeline Parallelism|@RUAN-ZX
Done|Zero1|@misstek
Done|Zero2|@misstek
Done|Zero3|@misstek
Done|Activation Checkpointing|@CurryRice233
Done|[Fused Adam](https://github.com/microsoft/DeepSpeed/pull/4343)|@CurryRice233
Done|Mixture of Experts (MoE)|@wangshuai09
Done|[RLHF](https://github.com/microsoft/DeepSpeed/pull/4343)|@wangshuai09 @CurryRice233
Done|[ZeRO Offload](https://github.com/microsoft/DeepSpeed/pull/4568)|@hipudding @CurryRice233
Done|[ZeRO Infinity](https://github.com/microsoft/DeepSpeed/pull/4809)|@misstek
Done|[1-bit Adam](https://github.com/microsoft/DeepSpeed/pull/4733)|@RUAN-ZX
Done|[1-bit LAMB](https://github.com/microsoft/DeepSpeed/pull/4733)|@RUAN-ZX
Done|[0/1 Adam](https://github.com/microsoft/DeepSpeed/pull/4733)|@minchao-sun
Done|Curriculum Learning|@minchao-sun
Done|Layer Dropping|@minchao-sun

**Comment by delock** [comment]
I see npu FusedAdam is implemented with `torch_npu.npu_apply_adam_w`.  In the future when implement new features, does NPU intend to support through torch_npu, or may also implement kernel in DeepSpeed as well?

**Comment by CurryRice233** [comment]
> I see npu FusedAdam is implemented with `torch_npu.npu_apply_adam_w`. In the future when implement new features, does NPU intend to support through torch_npu, or may also implement kernel in DeepSpeed as well?

The NPU supports two modes. Personally, I prefer the first one, where users can directly invoke the interface, regardless of the implementation

**Comment by realgump** [comment]
Hello, is there any plan to support FusedLamb on NPU?

**Comment by hipudding** [comment]
> Hello, is there any plan to support FusedLamb on NPU?

Honestly, support for this feature is not currently in our plan, but we will keep an eye on it. Thank you.
